File: go/modelalive/stable_test.go
package modelalive_test

import (
	"testing"

	"github.com/Ahaozandaburada/modelalive/go/modelalive"
)

func TestStablePrompts(t *testing.T) {
	prompts, err := modelalive.ListStablePrompts()
	if err != nil {
		t.Fatal(err)
	}
	if len(prompts) < 5 {
		t.Fatalf("expected >=5 prompts, got %d", len(prompts))
	}
}

func TestStableCompareIdentical(t *testing.T) {
	prompts, _ := modelalive.ListStablePrompts()
	responses := map[string][]string{}
	for _, p := range prompts {
		responses[p.ID] = []string{"fixture"}
	}
	base, err := modelalive.FingerprintFromResponses("gpt-4o", responses, nil, 1)
	if err != nil {
		t.Fatal(err)
	}
	cur, err := modelalive.FingerprintFromResponses("gpt-4o", responses, nil, 1)
	if err != nil {
		t.Fatal(err)
	}
	report := modelalive.CompareFingerprints(base, cur, 0.25)
	if !report.Stable {
		t.Fatalf("expected stable, mean=%f", report.MeanDistance)
	}
}

func TestStableDetectDrift(t *testing.T) {
	prompts, _ := modelalive.ListStablePrompts()
	baseMap := map[string][]string{}
	for _, p := range prompts {
		baseMap[p.ID] = []string{"alpha beta gamma"}
	}
	shifted := map[string][]string{}
	for k, v := range baseMap {
		shifted[k] = append([]string(nil), v...)
	}
	shifted["math_fixed"] = []string{"totally different output here now"}

	base, _ := modelalive.FingerprintFromResponses("gpt-4o", baseMap, nil, 1)
	cur, _ := modelalive.FingerprintFromResponses("gpt-4o", shifted, nil, 1)
	report := modelalive.CompareFingerprints(base, cur, 0.01)
	if report.Stable {
		t.Fatal("expected drift")
	}
}

func TestResponseDistanceZero(t *testing.T) {
	if modelalive.ResponseDistance("same", "same") != 0 {
		t.Fatal("expected zero distance")
	}
}

File: js/src/stable.test.ts
import assert from "node:assert/strict";
import {
  compareFingerprints,
  fingerprintFromResponses,
  listStablePrompts,
  responseDistance,
} from "./stable.js";

assert.ok(listStablePrompts().length >= 5);
assert.equal(responseDistance("hello world", "hello world"), 0);
assert.ok(responseDistance("hello world", "goodbye moon") > 0.2);

const responses = Object.fromEntries(listStablePrompts().map((p) => [p.id, ["fixture-response"]]));
const base = fingerprintFromResponses("gpt-4o", responses);
const same = fingerprintFromResponses("gpt-4o", responses);
const report = compareFingerprints(base, same, 0.25);
assert.equal(report.stable, true);

const shifted = { ...responses, math_fixed: ["999999"] };
const drift = compareFingerprints(base, fingerprintFromResponses("gpt-4o", shifted), 0.01);
assert.equal(drift.stable, false);

console.log("stable.test.ts ok");
